### DISCUSSION TITLE
Decreased processing speed when updating in versions 1.9.0 (and latest)

### DISCUSSION THREAD

**Discussion by azerotla2** [akvorado/akvorado#988] [Q&A]
Hi,

I would like to draw your attention to a significant slowdown in the performance of the service after updating to version 1.9.0 (and testing versions 1.9.1 and 1.9.2). I have observed an almost twofold reduction in performance. To analyze the situation, I implemented additional monitoring for ClickHouse and Kafka, focusing on three key graphs:
1. Increase in the number of records in ClickHouse (flows/sec)
2. Delay for the last record in ClickHouse (delay/seconds)
3. CPU utilization

In the attached graphs, you can see the difference in the service's performance before and after the update. It is evident that after the update, the lagging starts to increase. Later, it was discovered that the system, after the update, reaches the maximum processing of records at approximately 80k flows/sec. Whereas with the same configuration, version 1.8.3 processed 150k flows/sec.

![before_update](https://github.com/akvorado/akvorado/assets/84681368/f5b20809-d233-4fc6-bcd7-0a76c23bacca)
![after_update](https://github.com/akvorado/akvorado/assets/84681368/e0a74d2f-983c-49e3-9a1b-211dbed9d15a)

It is important to note that I did not make changes to the configuration files, except for replacing docker-compose files, and making changes to the .env file. I also rewrote lines in inlet.yaml according to the following rules:

`metadata:
  provider:
    type: snmp
    communities:
      ::/0: public`

After experimenting with the number of workers on ClickHouse and Kafka, I managed to increase the number of flows/sec to 120k, but this significantly increased CPU utilization (visible on the graph).
![add_workers](https://github.com/akvorado/akvorado/assets/84681368/56c82760-fa9f-4d46-a747-0b6edbb8b4bf)

Subsequently, rolling back the system to the previous number of workers allowed the system to process the accumulated queue at a speed of 180k flows/sec, and it is now working stably with normal CPU usage.
![rollback](https://github.com/akvorado/akvorado/assets/84681368/67e79854-6cfb-40eb-9582-72ac4c7569d7)

Is it possible that I missed some important configuration files, or does the new version really work twice as slowly? My observations indicate a slowdown specifically in the traffic enrichment speed.

I would appreciate your attention to this issue and am ready to provide additional details if necessary. This problem significantly affects the overall performance of the system, and I am confident that your team will find the best solution.

Thank you

**Comment by vincentbernat** [comment]
When rolling back to 1.8.3, you keep everything else as is, including Kafka/Clickhouse versions?

**Reply by vincentbernat** [reply]
If you are able to reproduce the issue easily, it would be very helpful to bisect between 1.8.3 and 1.9.0. There are 284 commits, you should be able to identify the culprit in about 9 steps.

**Reply by vincentbernat** [reply]
Also, which geoip provider is being used in 1.8.3 and 1.9.0. The default changed, with 1.8.3 using Maxmind if configured (and nothing if not) and IPInfo being used by default in 1.9.0.

**Reply by ghost** [reply]
We stayed with maxmind with the 1.8.3 to 1.9.2 change.

**Reply by vincentbernat** [reply]
Are you using sflow or netflow?

**Reply by ghost** [reply]
both :blush: 
ipfix for juniper and sflow for arista

**Reply by atiide** [reply]
I don't have any logs, but last time I used akvorado it could handle ~10k flows easy on my VM. I was using the latest build when metadata was introduced to inlet.yaml. Should be this commit https://github.com/akvorado/akvorado/commit/a86a1870515d1a3638f0259c65fb952434030c3f

Now  with the new build. The same VM can't handle the same flowrate. I have tried both IPinfo and Maxmind for enriching. Same result.

**Reply by vincentbernat** [reply]
This commit is here to convert the configuration from the previous format to the new format. The commit introducing the metadata provider is 7e3ca930ad4efb1de2ae3ff5ce4bd509cfd4b7a3. You can try to build 8064461ba47efdb358447831fd72bdc4b5a94bfe (`docker build -t ghcr.io/akvorado/akvorado:latest .`) and check the performance you get with it (don't forget to update your `docker-compose.yml` to use `:latest`). Then, try with 2980eb491c05331d96243be237a12921e72913fa. If you can see the different in performance between these two, it would be helpful.

**Reply by atiide** [reply]
I couldn't get https://github.com/akvorado/akvorado/commit/8064461ba47efdb358447831fd72bdc4b5a94bfe working.(Validator had that panic slice error). 
https://github.com/akvorado/akvorado/commit/b464a00be17f80fc09c7c5c2f82af049c00d4630 I used this instead.  
12h of running on the old build
```
cat /proc/loadavg 
3.97 3.32 3.00 5/983 3493018
```
I remember the newest build (I use :main) had load average of 6-7. Do you want me to grab any specific metrics from the old build?

**Reply by vincentbernat** [reply]
So, that's not the metadata change. I fear this may be a dependency update. While this is annoying, you can bisect from where you are:
```
git bisect start
git bisect good b464a00
git bisect bad v1.9.0
docker build -t ghcr.io/akvorado/akvorado:main .
# try the build
git bisect good # if load average is ok
git bisect bad # if not
```
Once the culprit commit is indentified, this should be easier. This should take about 6 steps.

An alternative would be to take a CPU profile on the good build and one on the bad build. Maybe we can identify the problem here. See https://demo.akvorado.net/docs/troubleshooting#profiling.

Do you use the routing component?

**Reply by atiide** [reply]
```
cat /proc/loadavg 
6.07 6.05 5.48 5/1114 305665
```

Not really that familiar with git bisecting so I went with the CPU profiling. Don't use any routing component. Only Maxmind for geoip. For my setup the static mapping was necessary. Both use identical configuration files. 
b464a00
![b464a00_profile](https://github.com/akvorado/akvorado/assets/137687151/6c3cb0e6-ee4b-4294-9bdb-9816f8e7cfa4)
![b464a00_heap](https://github.com/akvorado/akvorado/assets/137687151/0bc6ce6c-dc86-4f72-8761-4f9c5a7526eb)
aa5e5a4
![aa5e5a4_profile](https://github.com/akvorado/akvorado/assets/137687151/f69db432-3514-4575-b68c-accd692746ff)
![aa5e5a4_heap](https://github.com/akvorado/akvorado/assets/137687151/3086c9f7-2864-4a7b-90ee-64cb683c31b7)

I will let the newer build run, until flows per second drops to 0(that's how I noticed the system can't keep up with the flows). Then I will post another set of debugging svg if necessary.

**Reply by vincentbernat** [reply]
Thanks, that's very helpful! The regression is when the second routing provider was introduced. Can you try latest image once it is built? I think it should fix your issue. A profile would be nice to confirm the problem is really gone.

**Reply by atiide** [reply]
Running this build 73c286c
```
 cat /proc/loadavg 
4.86 5.18 4.82 9/1103 816452
```

![73c286c_heap](https://github.com/akvorado/akvorado/assets/137687151/c4154b38-af8c-47b1-b300-28fc06d9f705)
![73c286c_profile](https://github.com/akvorado/akvorado/assets/137687151/54e4542b-8639-44a7-8215-d78d3e68a331)

**Reply by vincentbernat** [reply]
While the load is higher (maybe more flows now than earlier in the day?), the profile shows it is better than the "good" version.

- Before: 14.61s on CPU for 30s
- After: 11.65s on CPU for 30s

I was expecting a slight improvement (as there is now a shortcut if you don't use the routing component). In term of number of flows/s, how does it feel?

**Reply by atiide** [reply]
I didn't really pay attention to the CPU load before. I noticed my issue when it would just not be able to process flows fast enough. 

From my perspective it seems to be ok now. I will see tomorrow how well it handles the extra load, because we monitor mainly office connections.

**Reply by atiide** [reply]
```
 cat /proc/loadavg
8.02 7.96 7.96 5/1152 1910718
```
![profile012](https://github.com/akvorado/akvorado/assets/137687151/e0b0c6c4-7d78-43f1-89a8-0d65dfedbadd)
![profile013](https://github.com/akvorado/akvorado/assets/137687151/5ca2efd3-1499-4d4c-ab73-0265a908fad6)

Still running into the same issue where at some point flows/s drops to 0. To be honest. I am not really sure if my VM is even powerful enough for this amount of flows so maybe I just need to allocate more resources

**Reply by vincentbernat** [reply]
While it is still possible to optimize (it seems the garbage collector is stealing a substantial amount of time, so we could find a way to do less allocations or reuse them), at some point, it is easier to throw hardware at the issue.

**Reply by vincentbernat** [reply]
For allocations, you can profile "allocs" (instead of "heap") to check what are the objects are often allocated and reallocated.

**Reply by atiide** [reply]
I ran into the same issue with the older "good" build. So my issues are actually caused by not enough resources. From my perspective the issue is resolved

**Comment by ghost** [comment]
I also have some performance problems, it shows as in https://github.com/akvorado/akvorado/discussions/561

> After experimenting with the number of workers on ClickHouse and Kafka, I managed to increase the number of flows/sec to 120k

which options did you tune @azerotla2 ?

I tried to downgrade 1.8.3  but clickhouse is failing, before debugging this I'd like to try throwing more ressources on the problem.

```
clickhouse:
  orchestrator-url: http://akvorado-orchestrator:8080
  kafka:
    consumers: 4
```
```
flow:
  inputs:
    - type: udp
      decoder: netflow
      listen: :2055
      workers: 6
      receive-buffer: 10485760
    - type: udp
      decoder: sflow
      listen: :6343
      workers: 6
      receive-buffer: 10485760
```
increasing the number of workers did also not help.
my config is basically the upstream one.

Help much appreciated! Thanks!

**Comment by toomuchio** [comment]
I've noticed this issue as well has any work been done to fix it, would it be advisable to upgrade to master @vincentbernat

**Reply by atiide** [reply]
One performance regression was fixed with https://github.com/akvorado/akvorado/commit/0fbdf13f27dda27d569afc5af7b0abee60e8f926

Use this in your compose file to get the latest built ghcr.io/akvorado/akvorado:main
